Utilities/ocvalidate/ValidateKernel.c

```diff
@@ -0,0 +1,440 @@
+  CONST CHAR8  *KextPlistPath;
+} KEXT_INFO;
+
+STATIC KEXT_PRECEDENCE mKextPrecedence[] = {
+  { "VirtualSMC.kext",    "Lilu.kext" },
+  { "WhateverGreen.kext", "Lilu.kext" },
+  //
+  // TODO: Add more kexts here...
+  //
+};
+STATIC UINTN mKextPrecedenceSize = ARRAY_SIZE (mKextPrecedence);
+
+STATIC KEXT_INFO mKextInfo[] = {
+  //
+  // NOTE: Index of Lilu should always be 0. Please add entries after this if necessary.
+  //
+  { "Lilu.kext",          "Contents/MacOS/Lilu",          "Contents/Info.plist" },
+  { "VirtualSMC.kext",    "Contents/MacOS/VirtualSMC",    "Contents/Info.plist" },
+  { "WhateverGreen.kext", "Contents/MacOS/WhateverGreen", "Contents/Info.plist" },
+  //
+  // TODO: Add more kexts here...
+  //
+};
+STATIC UINTN mKextInfoSize = ARRAY_SIZE (mKextInfo);
+
+UINT32
+CheckKernel (
+  IN  OC_GLOBAL_CONFIG  *Config
+  )
+{
+  UINT32              ErrorCount;
+  UINT32              Index;
+  OC_KERNEL_CONFIG    *UserKernel;
+  OC_PLATFORM_CONFIG  *UserPlatformInfo;
+  CONST CHAR8         *Arch;
+  CONST CHAR8         *BundlePath;
+  CONST CHAR8         *Comment;
+  CONST CHAR8         *ExecutablePath;
+  CONST CHAR8         *MaxKernel;
+  CONST CHAR8         *MinKernel;
+  CONST CHAR8         *PlistPath;
+  CONST CHAR8         *Identifier;
+  BOOLEAN             IsDisableLinkeditJettisonEnabled;
+  BOOLEAN             IsCustomSMBIOSGuidEnabled;
+  CONST CHAR8         *UpdateSMBIOSMode;
+  CONST CHAR8         *Base;
+  CONST UINT8         *Find;
+  UINT32              FindSize;
+  CONST UINT8         *Replace;
+  UINT32              ReplaceSize;
+  CONST UINT8         *Mask;
+  UINT32              MaskSize;
+  CONST UINT8         *ReplaceMask;
+  UINT32              ReplaceMaskSize;
+  CONST CHAR8         *KernelCache;
+  UINTN               IndexKextInfo;
+  UINTN               IndexKextPrecedence;
+  BOOLEAN             HasParent;
+  CONST CHAR8         *CurrentKext;
+  CONST CHAR8         *ParentKext;
+  CONST CHAR8         *ChildKext;
+
+  DEBUG ((DEBUG_VERBOSE, "config loaded into Kernel checker!\n"));
+  //
+  // Ensure Lilu to be always placed where it is supposed to be.
+  // 
+  ASSERT (AsciiStrCmp (mKextInfo[INDEX_KEXT_LILU].KextBundlePath, "Lilu.kext") == 0);
+  ASSERT (AsciiStrCmp (mKextInfo[INDEX_KEXT_LILU].KextExecutablePath, "Contents/MacOS/Lilu") == 0);
+  ASSERT (AsciiStrCmp (mKextInfo[INDEX_KEXT_LILU].KextPlistPath, "Contents/Info.plist") == 0);
+
+  ErrorCount                       = 0;
+  UserKernel                       = &Config->Kernel;
+  UserPlatformInfo                 = &Config->PlatformInfo;
+  IsDisableLinkeditJettisonEnabled = UserKernel->Quirks.DisableLinkeditJettison;
+  IsCustomSMBIOSGuidEnabled        = UserKernel->Quirks.CustomSmbiosGuid;
+  UpdateSMBIOSMode                 = OC_BLOB_GET (&UserPlatformInfo->UpdateSmbiosMode);
+  KernelCache                      = OC_BLOB_GET (&UserKernel->Scheme.KernelCache);
+
+  for (Index = 0; Index < UserKernel->Add.Count; ++Index) {
```

Comment:
Maybe add kernel driver and ACPI table duplicate search, just like you did with UEFI drivers?